### ISSUE TITLE
docs(notebooks): regression masterclass — pattern 2 (PyroxModule + pyrox_sample)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#76]
## Problem / Request

Port the same 7 regression models from the [Regression Masterclass design doc](design_docs/pyrox/examples/nn/regression_masterclass_eqx.md) to an executed `docs/notebooks/regression_masterclass_pyrox_sample.ipynb`, using **`pyrox._core.PyroxModule` + `pyrox_sample` / `pyrox_param`** to collapse the `eqx.tree_at` bridge into in-`__call__` self-registration. The boilerplate that path 1 puts in the model function moves *into* the module; the model function shrinks to `f = net(x); numpyro.sample("obs", dist.Normal(f, sigma), obs=y)`.

This is the **second of three sibling notebooks**. It directly builds on what pattern 1 teaches and shows what `pyrox._core` buys you.

## User Story

As a user who has read pattern 1 and now understands the `tree_at` bridge, I want to see the same 7 models written with `PyroxModule` so the pattern collapses: each layer registers its own priors during the forward pass, the model function loses its `numpyro.sample` cascade, and every existing layer in [`pyrox.nn._layers`](src/pyrox/nn/_layers.py) suddenly makes sense — they're all `PyroxModule` subclasses doing exactly this.

## Motivation

1. **Justify the abstraction.** `PyroxModule` already underpins all 12 layers in `pyrox.nn._layers` (`DenseReparameterization`, `DenseFlipout`, `MCDropout`, `RBFFourierFeatures`, etc.). Without a side-by-side comparison against pattern 1, users cannot see *why* it exists. This notebook is the rationale.
2. **Show the model-fn collapse.** Pattern 1's `numpyro` model functions average 15–25 lines of `numpyro.sample` + `eqx.tree_at` plumbing per model. Pattern 2's average ~5 lines: instantiate the layer, call it, attach the likelihood. The shrinkage is the teaching moment.
3. **Bridge to existing pyrox.nn.** After this notebook, users can read `DenseReparameterization.__call__` and immediately recognize the `pyrox_sample` calls as exactly the same pattern they just wrote.

## Proposed Structure

Same 7 models, same dataset, same metrics as the pattern 1 notebook — only the *module classes* and *model functions* change.

1. **Title + Colab badge** — `# Regression Masterclass — Pattern 2: PyroxModule + pyrox_sample`.
2. **Background** — what `PyroxModule` does (per-call context, fully-qualified site names, `pyrox_sample` vs `pyrox_param`); how it relates to pattern 1's `eqx.tree_at`.
3. **Setup + imports + dataset** — same as pattern 1.
4. **Model 1 — Bayesian linear** — `LinearRegressor(PyroxModule)` self-samples weights inside `__call__`.
5. **Model 2 — MAP MLP** — `MLP(PyroxModule)` with `pyrox_sample` for each weight + bias.
6. **Model 3 — MC-Dropout MLP** — Reuses [`pyrox.nn.MCDropout`](src/pyrox/nn/_layers.py) to show "the layer you'd write is already in the package".
7. **Model 4 — Bayesian MLP via NUTS** — Same `MLP(PyroxModule)`; only the inference call changes.
8. **Model 5 — SVR via RFF** — Reuses [`pyrox.nn.RBFFourierFeatures`](src/pyrox/nn/_layers.py); the user's job collapses to "wrap a `Linear` head".
9. **Model 6 — Approximate GP via hierarchical RFF** — Same RFF layer + a `pyrox_sample` for the amplitude in the user's wrapper.
10. **Model 7 — Deep RFF GP** — Compose two `RBFFourierFeatures` layers + a `Linear`.
11. **Side-by-side comparison cell** — table: pattern 1 LOC vs pattern 2 LOC for the model function only.
12. **Pointer to siblings** — pattern 1 (rawer) and pattern 3 (declarative).

**No helpers.** Same as pattern 1 — `MCMC` / `SVI` / `Predictive` are called directly; metric computation is inline.

## Mathematical Notes

The math is **identical** to pattern 1. What changes is the JAX/numpyro plumbing, not the model. The notebook should reuse the math blocks from the design doc verbatim and add one new block at the top:

### `pyrox_sample` semantics

When a `PyroxModule` method decorated with `@pyrox_method` is called, a per-instance context activates. Inside that context:

$$\texttt{self.pyrox\_sample(name, prior)} \;\equiv\; \texttt{numpyro.sample(f"\{scope\}.\{name\}", prior)}$$

with results memoized for the duration of the call (so two reads of `self.pyrox_sample("w", ...)` in the same `__call__` collapse to one numpyro site). The fully-qualified name `{ClassName}_{id}.{name}` keeps sibling instances of the same class distinct within a single trace.

This means a `PyroxModule` carrying $K$ random parameters declares $K$ numpyro sites *inside* `__call__`, with no `eqx.tree_at` needed — the module is its own placeholder. The model function becomes a one-liner:

```python
def model_fn(x, y=None):
    f = net(x)
    sigma = numpyro.sample("sigma", dist.HalfNormal(1.0))
    numpyro.sample("obs", dist.Normal(f, sigma), obs=y)
```

### MAP via `AutoDelta` still works unchanged

`numpyro.infer.SVI(model_fn, AutoDelta(model_fn), ...)` discovers all sites via tracing — including the `pyrox_sample` sites declared inside `__call__` — so the MAP path is identical to pattern 1 from the inference side.

## Reference Code

Pattern translation — `LinearRegressor`:

```python
# Pattern 1 — bare eqx.Module + tree_at in model fn (10+ lines model fn)
class LinearRegressor(eqx.Module):
    weight: jax.Array
    degree: int = eqx.field(static=True)
    def __init__(self, degree=1):
        self.degree = degree
        self.weight = jnp.zeros(degree + 1)
    def __call__(self, x):
        return einsum(self.features(x), self.weight, "n feat, feat -> n")

def model_linear(x, y=None, *, degree=1):
    net = LinearRegressor(degree=degree)
    w = numpyro.sample("w", dist.Normal(jnp.zeros(degree + 1), 1.0))
    net = eqx.tree_at(lambda m: m.weight, net, w)
    f = numpyro.deterministic("f", net(x))
    sigma = numpyro.sample("sigma", dist.HalfNormal(1.0))
    numpyro.sample("obs", dist.Normal(f, sigma), obs=y)
```

```python
# Pattern 2 — PyroxModule self-registers; model fn is ~3 lines
from pyrox._core import PyroxModule, pyrox_method

class LinearRegressor(PyroxModule):
    degree: int = eqx.field(static=True)

    def features(self, x):
        return jnp.stack([x ** p for p in range(self.degree + 1)], axis=-1)

    @pyrox_method
    def __call__(self, x):
        w = self.pyrox_sample("w", dist.Normal(jnp.zeros(self.degree + 1), 1.0))
        return einsum(self.features(x), w, "n feat, feat -> n")

def model_linear(x, y=None, *, degree=1):
    net = LinearRegressor(degree=degree)
    f = numpyro.deterministic("f", net(x))
    sigma = numpyro.sample("sigma", dist.HalfNormal(1.0))
    numpyro.sample("obs", dist.Normal(f, sigma), obs=y)
```

For Model 5 (SVR via RFF), the user doesn't even define a feature module — pyrox already ships [`RBFFourierFeatures`](src/pyrox/nn/_layers.py):

```python
from pyrox.nn import RBFFourierFeatures

class SVRRegressor(PyroxModule):
    rff: RBFFourierFeatures
    n_features: int = eqx.field(static=True)

    @pyrox_method
    def __call__(self, x):
        Phi = self.rff(x)                                   # (N, D)
        w = self.pyrox_sample("w", dist.Normal(jnp.zeros(self.n_features), 1.0))
        return einsum(Phi, w, "n d, d -> n")

def model_svr_rff(x, y=None, *, n_features=50, lengthscale=1.0):
    rff = RBFFourierFeatures(n_features=n_features, lengthscale=lengthscale, key=jr.PRNGKey(0))
    net = SVRRegressor(rff=rff, n_features=n_features)
    f = numpyro.deterministic("f", net(x))
    sigma = numpyro.sample("sigma", dist.HalfNormal(1.0))
    numpyro.sample("obs", dist.Normal(f, sigma), obs=y)
```

The MC-Dropout model gets even shorter — just instantiate [`pyrox.nn.MCDropout`](src/pyrox/nn/_layers.py) and chain it into the MLP composite.

## References

1. [`pyrox._core.PyroxModule`](src/pyrox/_core/pyrox_module.py) and [`pyrox_method`](src/pyrox/_core/pyrox_module.py) — the module + decorator this notebook teaches.
2. [`pyrox.nn._layers`](src/pyrox/nn/_layers.py) — the 12 layers already using this pattern; they become readable after this notebook.
3. Same model references as pattern 1: Aboleth, Rahimi & Recht 2007, Cutajar et al. 2017, Gal & Ghahramani 2016, Hoffman & Gelman 2014.
4. Phan, D., Pradhan, N., & Jankowiak, M. (2019). *Composable Effects for Flexible and Accelerated Probabilistic Programming in NumPyro*. [arXiv:1912.11554](https://arxiv.org/abs/1912.11554) — the trace mechanism that lets `pyrox_sample` work transparently from inside `__call__`.
5. Tran, D., Hoffman, M. D., et al. (2018). *Simple, Distributed, and Accelerated Probabilistic Programming* (Edward2). NeurIPS. [arXiv:1811.02091](https://arxiv.org/abs/1811.02091) — the prior art for "layers carry their own priors", which `PyroxModule` is the JAX/equinox-native version of.
6. Kidger, P. (2021). *equinox*. [Docs](https://docs.kidger.site/equinox/) — `PyroxModule` is a thin layer over `eqx.Module`.
7. [`design_docs/pyrox/examples/nn/regression_masterclass_eqx.md`](design_docs/pyrox/examples/nn/regression_masterclass_eqx.md) — same shared math source as patterns 1 and 3.

## Implementation Steps

- [ ] Author `docs/notebooks/regression_masterclass_pyrox_sample.py` in jupytext percent format.
- [ ] Header + Colab badge.
- [ ] Setup + imports (`from pyrox._core import PyroxModule, pyrox_method`; `from pyrox.nn import RBFFourierFeatures, MCDropout`) + dataset (inlined).
- [ ] One section per model — same as pattern 1, but each module subclasses `PyroxModule` and self-samples; each model fn is ≤ 4 lines.
- [ ] Models 5–7 reuse `pyrox.nn.RBFFourierFeatures`; Model 3 reuses `pyrox.nn.MCDropout`. Where the package layer fits exactly, *use it* and call attention to the line saved.
- [ ] Closing comparison cell: count LOC of each model fn in pattern 1 vs pattern 2; print a table.
- [ ] Convert + execute + commit `.ipynb` per docs workflow.
- [ ] `mkdocs.yml` nav entry adjacent to pattern 1.

## Definition of Done

- `docs/notebooks/regression_masterclass_pyrox_sample.ipynb` is committed with embedded outputs.
- Notebook executes end-to-end under `nbconvert --execute --inplace --timeout=600`.
- Same R² assertions as pattern 1 (≥ 0.85 for all non-degenerate models, ≥ 0.90 for Bayesian MLP).
- Each model fn body is ≤ 5 statements (cell assertion: `len(inspect.getsource(model_fn).strip().splitlines()) <= 8` after stripping decorator + signature).
- Models 3, 5, 6, 7 import the corresponding `pyrox.nn` layer rather than redefining it.

## Testing

- Same R² assertions as pattern 1.
- Additional cell asserts that the trace from `numpyro.handlers.trace(net.__call__)(x)` contains the expected `{ClassName}_*.{name}` sites — proves `pyrox_sample` registration works as documented.

## Documentation

- The notebook is the deliverable.
- Add a one-liner to [`docs/api/core.md`](docs/api/core.md) cross-linking to this notebook as the canonical "how do I use `PyroxModule`" example.

## Relationships

- **Parent epic:** [#36 — `[Epic] 4.C NN Uncertainty Patterns`](https://github.com/jejjohnson/pyrox/issues/36).
- **Sibling notebooks:**
  - Pattern 1 — `tree_at` raw equinox + numpyro.
  - Pattern 3 — `Parameterized` + `PyroxParam` declarative form.
- **Reuses (today):** [`pyrox._core.PyroxModule`](src/pyrox/_core/pyrox_module.py), [`pyrox.nn.RBFFourierFeatures`](src/pyrox/nn/_layers.py), [`pyrox.nn.MCDropout`](src/pyrox/nn/_layers.py).
- **Not blocked by** any other open issue.
